crates/bevy_ecs/src/query/access.rs

```diff
@@ -289,7 +318,7 @@ impl<T: SparseSetIndex> FilteredAccess<T> {
 
     pub fn extend_intersect_filter(&mut self, other: &FilteredAccess<T>) {
         self.without.intersect_with(&other.without);
-        self.with.intersect_with(&other.with);
+        self.with.extend_with_or(&other.with);
     }
```

Comment:
I'm kinda surprised that this is basically the only method I had to change to make tests work. Are there any other things that we have to consider?

Still not sure that the implementation is correct. I'll appreciate it if anyone can suggest other scenarios to cover with tests.